File: api_python_postgres/app/controllers/userController.py
import bcrypt
from fastapi import HTTPException
from app.config.database import get_pool
from app.utils.response import api_response
from app.schemas.user import UserRegister, UserLogin


# ========================================
# REGISTRAR USUARIO
# ========================================
async def register(payload: UserRegister):
    pool = get_pool()

    # VALIDAR SI EXISTE
    usuario_existe = await pool.fetchrow(
        "SELECT * FROM usuarios WHERE correo = $1", payload.correo
    )

    if usuario_existe:
        raise HTTPException(
            status_code=400,
            detail=api_response(False, "Correo ya registrado", None, "DUPLICATE_EMAIL"),
        )

    # HASH PASSWORD
    password_hash = bcrypt.hashpw(payload.password.encode(), bcrypt.gensalt()).decode()

    # INSERT SQL
    resultado = await pool.fetchrow(
        "INSERT INTO usuarios (nombre, correo, password) VALUES ($1, $2, $3) RETURNING id",
        payload.nombre,
        payload.correo,
        password_hash,
    )

    return api_response(
        True,
        "Usuario registrado correctamente",
        {"id": resultado["id"], "nombre": payload.nombre, "correo": payload.correo},
    )


# ========================================
# LOGIN
# ========================================
async def login(payload: UserLogin):
    pool = get_pool()

    usuario = await pool.fetchrow(
        "SELECT * FROM usuarios WHERE correo = $1", payload.correo
    )

    if not usuario:
        raise HTTPException(
            status_code=404,
            detail=api_response(False, "Usuario no encontrado"),
        )

    # VALIDAR PASSWORD
    password_correcto = bcrypt.checkpw(
        payload.password.encode(), usuario["password"].encode()
    )

    if not password_correcto:
        raise HTTPException(
            status_code=401,
            detail=api_response(False, "Contraseña incorrecta"),
        )

    return api_response(
        True,
        "Login exitoso",
        {"id": usuario["id"], "nombre": usuario["nombre"], "correo": usuario["correo"]},
    )


# ========================================
# LISTAR USUARIOS
# ========================================
async def get_users():
    pool = get_pool()

    result = await pool.fetch("SELECT id, nombre, correo FROM usuarios")

    return api_response(True, "Lista usuarios", [dict(r) for r in result])


# ========================================
# ELIMINAR USUARIO
# ========================================
async def delete_user(user_id: int):
    pool = get_pool()

    await pool.execute("DELETE FROM usuarios WHERE id = $1", user_id)

    return api_response(True, "Usuario eliminado")

File: api_node_mysql/src/controllers/userController.js
const pool = require('../db');
const bcrypt = require('bcrypt');

// Respuesta unificada
const apiResponse = (success, message, data = null, error = null) => {
    return { success, message, data, error };
};

// Obtener todos los usuarios
const getUsuarios = async (req, res) => {
    try {
        const [rows] = await pool.query('SELECT id, nombre, correo, edad, created_at FROM usuarios ORDER BY id');
        res.json(apiResponse(true, 'Usuarios obtenidos correctamente', rows));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error al obtener usuarios', null, error.message));
    }
};

// Obtener un usuario por ID
const getUsuarioById = async (req, res) => {
    const { id } = req.params;
    try {
        const [rows] = await pool.query('SELECT id, nombre, correo, edad, created_at FROM usuarios WHERE id = ?', [id]);
        if (rows.length === 0) {
            return res.status(404).json(apiResponse(false, 'Usuario no encontrado', null, 'USER_NOT_FOUND'));
        }
        res.json(apiResponse(true, 'Usuario obtenido correctamente', rows[0]));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error al obtener el usuario', null, error.message));
    }
};

// Crear un nuevo usuario (registro)
const createUsuario = async (req, res) => {
    const { nombre, correo, password, edad } = req.body;
    
    // Validar campos requeridos
    if (!nombre || !correo || !password) {
        return res.status(400).json(apiResponse(false, 'Nombre, correo y password son requeridos', null, 'MISSING_FIELDS'));
    }
    
    try {
        // Verificar si el correo ya existe
        const [existing] = await pool.query('SELECT * FROM usuarios WHERE correo = ?', [correo]);
        if (existing.length > 0) {
            return res.status(400).json(apiResponse(false, 'El correo ya está registrado', null, 'DUPLICATE_EMAIL'));
        }
        
        // Encriptar password
        const passwordHash = await bcrypt.hash(password, 10);
        
        // Insertar usuario
        const [result] = await pool.query(
            'INSERT INTO usuarios (nombre, correo, password, edad) VALUES (?, ?, ?, ?)',
            [nombre, correo, passwordHash, edad || null]
        );
        
        res.status(201).json(apiResponse(true, 'Usuario registrado correctamente', {
            id: result.insertId,
            nombre,
            correo,
            edad
        }));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error al crear el usuario', null, error.message));
    }
};

// Login de usuario
const loginUsuario = async (req, res) => {
    const { correo, password } = req.body;
    
    if (!correo || !password) {
        return res.status(400).json(apiResponse(false, 'Correo y password son requeridos', null, 'MISSING_FIELDS'));
    }
    
    try {
        const [rows] = await pool.query('SELECT * FROM usuarios WHERE correo = ?', [correo]);
        
        if (rows.length === 0) {
            return res.status(401).json(apiResponse(false, 'Credenciales incorrectas', null, 'INVALID_CREDENTIALS'));
        }
        
        const usuario = rows[0];
        const passwordValido = await bcrypt.compare(password, usuario.password);
        
        if (!passwordValido) {
            return res.status(401).json(apiResponse(false, 'Credenciales incorrectas', null, 'INVALID_CREDENTIALS'));
        }
        
        // Login exitoso (sin JWT por ahora)
        res.json(apiResponse(true, 'Login exitoso', {
            id: usuario.id,
            nombre: usuario.nombre,
            correo: usuario.correo,
            edad: usuario.edad
        }));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error en el login', null, error.message));
    }
};

// Actualizar un usuario
const updateUsuario = async (req, res) => {
    const { id } = req.params;
    const { nombre, correo, edad, password } = req.body;
    
    try {
        let query = 'UPDATE usuarios SET nombre = ?, correo = ?, edad = ?';
        let params = [nombre, correo, edad];
        
        if (password) {
            const passwordHash = await bcrypt.hash(password, 10);
            query += ', password = ?';
            params.push(passwordHash);
        }
        
        query += ' WHERE id = ?';
        params.push(id);
        
        const [result] = await pool.query(query, params);
        
        if (result.affectedRows === 0) {
            return res.status(404).json(apiResponse(false, 'Usuario no encontrado', null, 'USER_NOT_FOUND'));
        }
        
        res.json(apiResponse(true, 'Usuario actualizado correctamente'));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error al actualizar el usuario', null, error.message));
    }
};

// Eliminar un usuario
const deleteUsuario = async (req, res) => {
    const { id } = req.params;
    try {
        const [result] = await pool.query('DELETE FROM usuarios WHERE id = ?', [id]);
        
        if (result.affectedRows === 0) {
            return res.status(404).json(apiResponse(false, 'Usuario no encontrado', null, 'USER_NOT_FOUND'));
        }
        
        res.json(apiResponse(true, 'Usuario eliminado correctamente'));
    } catch (error) {
        console.error(error);
        res.status(500).json(apiResponse(false, 'Error al eliminar el usuario', null, error.message));
    }
};

module.exports = {
    getUsuarios,
    getUsuarioById,
    createUsuario,
    loginUsuario,
    updateUsuario,
    deleteUsuario
};